### PR TITLE
Adds costing to diafiltration superstructure flowsheet

### PR DESCRIPTION
**Pull request by mollydougher**
## Summary/Motivation:
Extends the existing diafiltration costing framework to the superstructure flowsheet contributed in #104. This branch is built from the work within #104 and #105 and thus should be merged after both of these PRs have been merged.

## Changes proposed in this PR:
- Apply the diafiltration cost model to @jas-yao 's superstructure flowsheet with precipitators.
- Employ costing objectives (minimize cost) instead of recovery objectives (maximize recovery) to the flowsheet.

## Reviewer's checklist / merge requirements:
- [ ] The head branch (i.e. the "source" of the changes) is not the `main` branch on the PR author's fork
- [ ] Documentation
- [ ] Tests
- [ ] Diagnostic tests for models

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.md file
   at the top level of this directory.

2. I represent I am authorized to make the contributions and grant the license. If my employer has
   rights to intellectual property that includes these contributions, I represent that I have
   received permission to make contributions and grant the required license on behalf of that
   employer.

### CODE REVIEW

**Comment by bpaul4** [comment]
While addressing #147, I noted that `src/prommis/nanofiltration/membrane_cascade_flowsheet/utils.py` has two instances of bare exceptions. It's best to either except a specific type of exception or use a Boolean type or attribute check instead.

These exceptions are not triggered by `solve_diafiltration.py`, but additionally neither module is executed by the tests. @mollydougher can you please use this pull request to add a test module to cover this script, and replace the bare exceptions?

**Comment by lbianchi-lbl** [comment]
@mollydougher do you have an estimate of how close to being merged this is? For the June release, we'd want all work to be completed by the end of this week.

**Comment by mollydougher** [comment]
> While addressing #147, I noted that `src/prommis/nanofiltration/membrane_cascade_flowsheet/utils.py` has two instances of bare exceptions. It's best to either except a specific type of exception or use a Boolean type or attribute check instead.
> 
> These exceptions are not triggered by `solve_diafiltration.py`, but additionally neither module is executed by the tests. @mollydougher can you please use this pull request to add a test module to cover this script, and replace the bare exceptions?

@jas-yao I replaced these bare exceptions in your `utils.py` file with `ValueError`s, but I am unsure of what the edge cases are that will raise these exceptions. Do you have any input on writing a test module for these exceptions?

File: src/prommis/nanofiltration/costing/diafiltration_cost_model.py [hunk 1]
```diff
@@ -214,7 +215,8 @@ def cost_membrane_pressure_drop(
         vol_flow_perm,
     ):
         """
-        Costing method for membrane pressure drop
+        Costing method for membrane pressure drop.
```
Comment (Lingyan90) [inline src/prommis/nanofiltration/costing/diafiltration_cost_model.py]: The current `cost_membrane_pressure_drop` function appears to compute utility cost attributable to membrane pressure drop. To make this explicit, please rename it to something like `cost_utility_membrane_pressure_drop` to differentiate from equipment cost.
Also, for the docstring at ~L218, consider:
“Costing method for the membrane’s utility cost due to pressure drop.”

File: src/prommis/nanofiltration/costing/diafiltration_cost_model.py [hunk 2]
```diff
@@ -295,154 +297,264 @@ def variable_operating_cost_constraint(blk):
...
-            domain=NonNegativeReals,
-            units=blk.costing_package.base_currency / blk.costing_package.base_period,
-            doc="Unit variable operating cost",
-        )
+            @blk.Constraint()
+            def capital_cost_constraint(blk):
+                return blk.capital_cost == units.convert(
+                    blk.pump_correlation_factor
+                    * (inlet_vol_flow * inlet_pressure) ** blk.pump_exponential_factor,
+                    to_units=blk.costing_package.base_currency,
```
Comment (Lingyan90) [inline src/prommis/nanofiltration/costing/diafiltration_cost_model.py]: The pump capital-cost equation used here differs from the traditional base-cost/size scaling form. In this equation, cost is expressed directly as a power law in duty (e.g., QΔP) with a fitted coefficient. Is this how this should be calculated based on your citation?

File: src/prommis/nanofiltration/costing/diafiltration_cost_model.py [hunk 3]
```diff
@@ -451,81 +563,114 @@ def cost_precipitator(
...
+                domain=NonNegativeReals,
+                units=blk.costing_package.base_currency,
+                doc="Unit capital cost",
+            )
+
+            @blk.Constraint()
+            def capital_cost_constraint(blk):
+                return blk.capital_cost == units.convert(
+                    blk.precipitator_factor_capital * precip_volume
+                    + 12000 * units.USD_1998,
```
[the remark this answers is not in the document]
Comment (mollydougher) [reply src/prommis/nanofiltration/costing/diafiltration_cost_model.py]: @jas-yao, can you remind me which pressure case you used to estimate the $12,000 base cost of the precipitators? 

@Lingyan90, I have also added a citation for Jason's most recent paper which has more detail on the simple costing model.

[the remark this answers is not in the document]
Comment (mollydougher) [reply src/prommis/nanofiltration/costing/diafiltration_cost_model.py]: I have revised this after discussing with @jas-yao. The previous price parameters of $12,000 and $1,000/m3 (which can be found in Jason's paper) are already price adjusted. Thus I "undid" the price adjustment in the most recent commit to stay in 1998 dollars as used in the reference plot. Both cost parameters come from the 150 psig line;  where $6,000 is the approximate minimum cost and $500/m3 is the approximate slope. I have also added comments explaining this.

File: src/prommis/nanofiltration/costing/tests/test_diafiltration_cost_model.py [hunk 4]
```diff
@@ -130,3 +130,111 @@ def test_diafiltration_costing():
...
+
+
+@pytest.mark.component
+def test_simple_costing():
+    # create a model
+    m = ConcreteModel()
+    m.fs = FlowsheetBlock(dynamic=False)
+    m.fs.costing = DiafiltrationCosting()
+
+    # define global parameters
```
[the remark this answers is not in the document]
Comment (mollydougher) [reply src/prommis/nanofiltration/costing/tests/test_diafiltration_cost_model.py]: Upon further consideration, I think it makes more sense for these to be defined as variables associated with specific `UnitModelBlocks` in the flowsheet, not global parameters.
